src/serialize.h

```diff
@@ -1114,4 +1178,74 @@ size_t GetSerializeSizeMany(int nVersion, const T&... t)
     return sc.size();
 }
 
+/** Wrapper that overrides the GetParams() function of a stream. */
+template<typename Params, typename Stream>
+class ParamsStream
+{
+    const Params& m_params;
+    Stream& m_substream;
+public:
+    ParamsStream(const Params& params, Stream& substream) : m_params(params), m_substream(substream) {}
+    template<typename U> inline ParamsStream& operator<<(const U& obj) { ::Serialize(*this, obj); return *this; }
+    template<typename U> inline ParamsStream& operator>>(U&& obj) { ::Unserialize(*this, obj); return *this; }
+    inline void write(const char* ptr, size_t size) { m_substream.write(ptr, size); }
+    inline void read(char* ptr, size_t size) { m_substream.read(ptr, size); }
+    inline size_t size() const { return m_substream.size(); }
+    inline const Params& GetParams() { return m_params; }
+    inline int GetVersion() const { return m_substream.GetVersion(); }
+    inline int GetType() const { return m_substream.GetType(); }
+    inline Stream& GetSubStream() const { return m_substream; }
+};
+
+/** Wrapper that serializes objects with the specified parameters. */
+template<typename Params, typename T>
+class ParamsWrapper
+{
+    static_assert(std::is_lvalue_reference<T>::value, "ParamsWrapper needs an lvalue reference type T");
+    const Params& m_params;
+    T m_object;
+
+public:
+    explicit ParamsWrapper(const Params& params, T obj) : m_params(params), m_object(obj) {}
+
+    //! Serialize to another ParamsStream: optimize by skipping it.
```

Comment:
This may be because I don't have a good understanding of serialize.h in general so perhaps you can just ignore this, but I don't understand what's being skipped here.

Reply:
It's an attempt at avoiding crazy template expansion in the compiler when multiple levels of serializers each use `WithParams` (the Stream object that would hit the bottom would be ParamsStream<ParamsInner, ParamsStream<ParamsMid, ParamsStream<ParamsOuter, BaseStream>>>, and all its methods need instantiating). This optimization will just result in `ParamsStream<ParamsInner, BaseStream>`.